[fastlane_core] add clearer error message when incorrectly defining tests that need xcodebuild

revolter opened the pull request:
<!-- Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->

### Checklist
- [ ] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->

This is to prevent issues like: https://circleci.com/gh/fastlane/fastlane/36544

<!-- If it fixes an open issue, please link to the issue here. -->

### Description
<!-- Describe your changes in detail. -->
<!-- Please describe in detail how you tested your changes. -->

<img width="1247" alt="image" src="https://user-images.githubusercontent.com/5748627/66717592-689d2900-ede3-11e9-8e73-377eeafaf389.png">

### Testing Steps
<!-- Optional: steps, commands, or code used to test your changes. -->
<!-- Providing these will reduce the time needed for testing and review by the fastlane team. -->

revolter committed:
Add clearer error message when incorrectly defining tests that need Xcode

revolter commented:
But, I now noticed that we have tests running on macOS with Xcode installed, so, why this parameter?

fastlane-bot-helper commented:
<!--
  0 Errors
  1 Warning: Pull Request is Work in Progre...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Pull Request is Work in Progress</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

janpio reviewed (changes requested):
`requires_xcodebuild: true` is to being able to disable test on Windows and Linux that require Xcode to be installed, often via the Xcode version check you also changed. 

If you get these on macOS, something else is broken.

revolter commented:
Hmm, I didn't pay much attention to the OS, and maybe I incorrectly thought that they failed on macOS.

revolter committed:
Make it clear that the code line is an example

And not the actual test that failed

janpio commented:
The GH run you linked to was Ubuntu, so it would make sense.

Did I understand correctly that you added additional tests, and did not set `requires_xcodebuild` thus giving you this error message? Then you changes may be useful after all - although in theory this would have to be output at _all_ locations where we use Xcode/`xcodebuild`. Not optimal.

Btw: I think `require_xcode` and `require_xcodebuild` might be different things, you have to check the codebase.

revolter commented:
> The GH run you linked to was Ubuntu, so it would make sense.

Exactly. So https://github.com/fastlane/fastlane/pull/15475#issuecomment-541015559 came out of confusion, so it should be ignored.

> Did I understand correctly that you added additional tests, and did not set `requires_xcodebuild` thus giving you this error message?

Yes, and it took me quite a while to figure it out. Hence, this change.

> although in theory this would have to be output at all locations where we use Xcode/`xcodebuild`. Not optimal.

What do you mean by this? If _x_ tests that indirectly use `xcodebuild`, don't specify that parameter, the error message that I added **will** be output _x_ times.

> Btw: I think `require_xcode` and `require_xcodebuild` might be different things, you have to check the codebase.

Meaning that the PR title is a bit misleading? _(I updated it now ⬆️)_

janpio commented:
> What do you mean by this? If _x_ tests that indirectly use `xcodebuild`, don't specify that parameter, the error message that I added **will** be output _x_ times.
> 
> >

Only if this exact code is executed. There are x other locations that _also_ use `xcodebuild` or Xcode and which do not output this error message.

joshdholtz commented on fastlane_core/lib/fastlane_core/helper.rb:
```diff
@@ -163,7 +163,12 @@ def self.xcode_version
     # @return true if Xcode version is higher than 8.3
     def self.xcode_at_least?(version)
-      FastlaneCore::UI.user_error!("Unable to locate Xcode. Please make sure to have Xcode installed on your machine") if xcode_version.nil?
+      if self.ci?
```
Do we actually want to make this `self.mac?`? 🤔  Since this would be nice to see on any linux/windows run and not necessarily just on CI?

janpio replied:
Not really, this is explicit communication to a developer. As I mentioned above, this is only 1 location out of hundreds that use Xcode (although one of the most "popular" ones), so this is also not complete.

revolter replied:
Why not? Maybe a Linux user clones the branch of someone using macOS that added some tests, but forgot to specify that flag. In which case, for that other person, the tests pas, but for the Linux user, they fail. And with this PR and Josh's change, he will knew exactly what the problem is, when running the tests locally, because the error message is clearer.

And back to your point, if xcodebuild is used from multiple places (which I'm not doubting), how does it work on Linux? And, we could fix it (e.g. make them all pass through a single function) in another PR, right? Meaning, if this is ok, it could be merged, right?

joshdholtz reviewed (commented):
Love this change! Just one quick question on the logic 😇

revolter commented:
I don't think that this is needed anymore, hence, closing it.
